src/manifolds/StiefelCanonicalMetric.jl

```diff
@@ -0,0 +1,115 @@
+@doc raw"""
+    CanonicalMetric <: Metric
+
+The Canonical Metric is name used on several manifolds, for example for the [`Stiefel`](@ref)
+manifold, see for example[^EdelmanAriasSmith1998].
+
+[^EdelmanAriasSmith1998]:
+    > Edelman, A., Ariar, T. A., Smith, S. T.:
+    > _The Geometry of Algorihthms with Orthogonality Constraints_,
+    > SIAM Journal on Matrix Analysis and Applications (20(2), pp. 303–353, 1998.
+    > doi: [10.1137/S0895479895290954](https://doi.org/10.1137/S0895479895290954)
+    > arxiv: [9806030](https://arxiv.org/abs/physics/9806030)
+"""
+struct CanonicalMetric <: RiemannianMetric end
+
+@doc raw"""
+    q = exp(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, CanonicalMetric}, p, X)
+    exp!(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, q, CanonicalMetric}, p, X)
+
+Compute the exponential map on the [`Stiefel`](@ref)`(n,k)` manifold with respect to the [`CanonicalMetric`](@ref).
+
+First, decompose The tangent vector ``X`` into its horizontal and vertical component with
+respect to ``p``, i.e.
+
+```math
+X = pp^{\mathrm{T}}X + (I_n-pp^{\mathrm{T}})X,
+```
+where ``I_n`` is the ``n\times n`` identity matrix.
+We introduce ``A=p^{\mathrm{T}}X`` and ``QR = (I_n-pp^{\mathrm{T}})X``` the qr decomposition
+of the vertical component. Then using the matrix exponential ``operatorname{Exp}`` we introduce ``B`` and ``C`` as
+
+```math
+\begin{pmatrix}
+B\\C
+\end{pmatrix}
+\coloneqq
+\operatorname{Exp}\left(
+\begin{pmatrix}
+A & -R^{\mathrm{T}}\\ R & 0
+\end{pmatrix}
+\right)
+\begin{pmatrix}I_k\\0\end{pmatrix}
+```
+
+the exponential map reads
+
+```math
+q = \exp_p X = pC + QB.
+```
+For more details, see [^EdelmanAriasSmith1998][^Zimmermann2017].
+"""
+exp(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, ::Any...) where {n,k}
+
+function exp!(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, Q, p, X) where {n,k}
+    A = p' * X
+    QR = qr(X - p * A)
+    BC_ext = exp([A -QR.R'; QR.R, zeros(k, k)])
+    q .= [p Matrix(QR.Q)] * BC_ext[:, 1:k]
+    return q
+end
+
+@doc raw"""
+    inner(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, X, CanonicalMetric}, p,, X, Y)
+
+compute the inner procuct on the [`Stiefel`](@ref) manifold with respect to the
+[`CanonicalMetric`](@ref). The formula reads
+
+```
+g_p(X,Y) = \operatorname{tr}\bigl( X^{\mathrm{T}}(I_n - \frac{1}{2}pp^{\mathrm{T}})Y \bigr).
+```
+"""
+function inner(::MetricManifold{ℝ,Stiefel{n,k,ℝ},X,CanonicalMetric}, p, X, Y) where {n,k}
+    return tr(X' * (Matrix(I, n, n) - 0.5 * p * p') * Y)
```

Comment:
What about something like this?
```suggestion
    return dot(X, (Y .- (p * (p'Y)) ./ 2))
```